### PR TITLE
add etcd into open source DC/OS

### PR DESCRIPTION
**Pull request by mainred**
## High-level description

This PR integrates ETCD into open source DC/OS.

It's basically a copy and paste from  [ETCD in EE DC/OS PR](https://github.com/mesosphere/dcos-enterprise/pull/6705)

## Corresponding DC/OS tickets (required)

  - [DCOS-59004](https://jira.mesosphere.com/browse/DCOS-59004) Port POC code from EE to Open.

## Related tickets (optional)

  - [DCOS-ID](https://jira.mesosphere.com/browse/DCOS-<number>) JIRA title / short description.

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain.
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]

### CODE REVIEW

File: packages/etcd/extra/etcd_discovery/etcd_discovery.py [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+
+    def _is_node_healthy(self, node: str) -> bool:
+        result = self._execute_etcdctl(node, ["endpoint", "health"])
+        healthy = result.returncode == 0
+        return healthy
+
+    def _execute_etcdctl(self, endpoint_ip: str,
+                         args: List[str]) -> subprocess.CompletedProcess:
+        cmd = [
+            self._etcdctl_path,
```
[the remark this answers is not in the document]
Comment (mainred) [reply packages/etcd/extra/etcd_discovery/etcd_discovery.py]: Yes, we can. I will take a note, and introduce this path in the following EE PR.

File: packages/adminrouter/extra/src/includes/http/master.conf [hunk 2]
```diff
@@ -53,3 +53,9 @@ upstream dcos_net {
 upstream dcos_ui_update_service {
     server unix:/run/dcos/dcos-ui-update-service.sock;
 }
+
+# Name: etcd
```
[the remark this answers is not in the document]
Comment (mainred) [reply packages/adminrouter/extra/src/includes/http/master.conf]: Done, removed

File: packages/windows.treeinfo.json [hunk 3]
```diff
@@ -38,6 +38,7 @@
     "dnspython",
     "docker-gc",
     "dvdcli",
+    "etcd",
```
[the remark this answers is not in the document]
Comment (mainred) [reply packages/windows.treeinfo.json]: This indicates whether or not a package should be built for windows.

**Comment by mainred** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by mainred** [comment]
@mesosphere-mergebot request-review  @vespian  @ichernetsky

**Comment by mainred** [comment]
Hey @travisvan  and @vespian , could you please take a look?

File: packages/etcd/extra/systemd/etcd.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+     --name "etcd-${IP_PRIVATE}" \
+     --data-dir /var/lib/dcos/etcd/default.etcd/ \
+     --listen-peer-urls "http://${IP_PRIVATE}:2380" \
+     --listen-client-urls "http://${IP_PRIVATE}:2379,http://localhost:2379" \
+     --initial-advertise-peer-urls "http://${IP_PRIVATE}:2380" \
+     --initial-cluster $CLUSTER_NODES \
+     --initial-cluster-state $CLUSTER_STATE \
+     --initial-cluster-token "etcd-dcos" \
+     --advertise-client-urls "http://${IP_PRIVATE}:2379,http://localhost:2379" \
+     --enable-v2=true
```
[the remark this answers is not in the document]
Comment (mainred) [reply packages/etcd/extra/systemd/etcd.sh]: It's a good point.

File: packages/etcd/extra/etcd_discovery/etcd_discovery.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mainred) [reply packages/etcd/extra/etcd_discovery/etcd_discovery.py]: I have updated etcd_discover.py to support both secure and insecure mode.
EE PR reusing OSS code can be found here https://github.com/mesosphere/dcos-enterprise/pull/6926

**Comment by mainred** [comment]
Previous CI is green, a new force push is made to rebase from the master branch.

File: packages/etcd/extra/etcd_discovery/etcd_discovery.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vespian) [reply packages/etcd/extra/etcd_discovery/etcd_discovery.py]: @mainred Do we have a Jira tracking this effort? Can't find it in the epic (https://jira.mesosphere.com/browse/DCOS-58411)

File: packages/etcd/extra/etcd_discovery/etcd_discovery.py [hunk 5]
```diff
@@ -0,0 +1,641 @@
...
+# Location of the detect IP address detection script
+DETECT_IP_SCRIPT = '/opt/mesosphere/bin/detect_ip'
+
+
+# In theory we could use the dcos_internal_utils.utils library, but in practice
+# the amount of deps it pulls in is huge compared to just a simple `execute
+# script and verify the output` function that we want
+def detect_ip() -> str:
+    machine_ip = subprocess.check_output(
+        [DETECT_IP_SCRIPT], stderr=subprocess.DEVNULL).decode('ascii').strip()
```
Comment (timaa2k) [inline packages/etcd/extra/etcd_discovery/etcd_discovery.py]: Can we try to not hide any errors here? `stderr=subprocess.STDOUT` may do the trick.

File: packages/etcd/extra/etcd_discovery/etcd_discovery.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mainred) [reply packages/etcd/extra/etcd_discovery/etcd_discovery.py]: https://jira.mesosphere.com/browse/DCOS-61012

File: packages/etcd/extra/etcd_discovery/etcd_discovery.py (hunk 5 above)
Comment (mainred) [reply packages/etcd/extra/etcd_discovery/etcd_discovery.py]: Done.

**Comment by mainred** [comment]
@vespian @jkoelker @timaa2k , just added a minor change to capture the stderr of `detect_ip` through subprocess.stdout. May I have your another approval, please.

**Pull request merged**